src/libAtomVM/ets.c

```diff
@@ -388,3 +420,117 @@ EtsErrorCode ets_delete(term ref, term key, term *ret, Context *ctx)
 
     return EtsOk;
 }
+
+bool operation_to_tuple4(term operation, term *position, term *increment, term *threshold, term *set_value)
+{
+    if (term_is_integer(operation)) {
+        *increment = operation;
+        *position = term_from_int(2);
+        *threshold = term_invalid_term();
+        *set_value = term_invalid_term();
+        return true;
+    }
+
+    if (!term_is_tuple(operation)) {
+        return false;
+    }
+    int n = term_get_tuple_arity(operation);
+    if (n != 2 && n != 4) {
+        return false;
+    }
+
+    term pos = term_get_tuple_element(operation, 0);
+    term incr = term_get_tuple_element(operation, 1);
+    if (!term_is_integer(pos) || !term_is_integer(incr)) {
+        return false;
+    }
+
+    if (n == 2) {
+        *position = pos;
+        *increment = incr;
+        *threshold = term_invalid_term();
+        *set_value = term_invalid_term();
+        return true;
+    }
+
+    term tresh = term_get_tuple_element(operation, 2);
+    term set_val = term_get_tuple_element(operation, 3);
+    if (!term_is_integer(tresh) || !term_is_integer(set_val)) {
+        return false;
+    }
+
+    *position = pos;
+    *increment = incr;
+    *threshold = tresh;
+    *set_value = set_val;
+    return true;
+}
+
+EtsErrorCode ets_update_counter(term ref, term key, term operation, term default_value, term *ret, Context *ctx)
+{
+    struct EtsTable *ets_table = term_is_atom(ref) ? ets_get_table_by_name(&ctx->global->ets, ref, TableAccessWrite) : ets_get_table_by_ref(&ctx->global->ets, term_to_ref_ticks(ref), TableAccessWrite);
+    if (ets_table == NULL) {
```

Comment:
Since the most likely outcome is a found table, I would do ` if (IS_NULL_PTR(ets_table)) {` (that is a short hand of `if (UNLIKELY(ets_table == NULL)) {`).